flink-runtime/src/main/java/org/apache/flink/runtime/leaderretrieval/ZooKeeperLeaderRetrievalDriver.java

```diff
@@ -69,8 +68,40 @@ public class ZooKeeperLeaderRetrievalDriver implements LeaderRetrievalDriver {
 
     private final FatalErrorHandler fatalErrorHandler;
 
+    /**
+     * Each {@code ZooKeeperLeaderRetrievalDriver} has its own watcher initialized. There is a bug
```

Comment:
> This change is meant as a draft and should rather be used as a base for a discussion whether [FLINK-33053](https://issues.apache.org/jira/browse/FLINK-33053) is actually an issue that should be addressed in Flink itself. As far as I understand, the thread leakage is only happening in test code when the ZooKeeper test server implementation is set up.

@KarmaGYZ Is it correct that the thread leakage happens in ZooKeeper's TestServer implementation? :thinking:

Reply:
Sorry, I'm not familiar with it. At least I can't find a public api to retrieve the watch info. Maybe you can verify with [this guide](https://kb.altinity.com/altinity-kb-setup-and-maintenance/altinity-kb-zookeeper/altinity-kb-how-to-check-the-list-of-watches/) to see whether we can get the watches of TestServer and wether there are leakage.

Reply:
I'm just asking because you mentioned in the description of FLINK-33053 that you observed the issue in stress tests. Did you use the MiniCluster and ZooKeeper in a single JVM or did you have your ZooKeeper deployed separately for the stress tests?

I'd like to understand whether it's on the ZooKeeper side (that's how it sounds to me right now) or on the Flink/Curator side. If I misunderstood the discussion in FLINK-33053 and the related PR #23415 and it's not a ZooKeeper server issue, fixing it on the Flink side is reasonable.

Reply:
No, I did the stress tests with a real ZK cluster. So, I'm afraid we might need an e2e test for it. 
The leackage is happened in the ZK side. Your fix loods reasonable to me, at least for the current HA mechanism.

Reply:
hm, ok ... thanks for clarification. But that brings me to the point where I would be in favor of not fixing it on Flink's side at all. That needs to be addressed in ZooKeeper (as @tisonkun pointed out with his reference to ZOOKEEPER-4625). What's your take on this?

Reply:
I think the core issue is a specific component should take the responsibility to manage the lifecycle of watches. This component can be ZooKeeper, Curator, or Flink. ZK can provide a "closeWatch" interface to reconcile watch references on the server-side. Similarly, Curator and Flink can proactively handle this task and clean up by using "closeAll," just like what you did in this PR.
I personally believe that let ZK to handle this task is more reasonable. However, considering the release cycles of ZK and the need to maintain compatibility with multiple versions of ZK in Flink, we may need to proactively manage the lifecycle of watches within Flink in the foreseeable future.